lib/pid/src/PID.cpp

```diff
@@ -1,492 +1,112 @@
-  prevControllerOutput_ *= (outMax - outMin) / outSpan_;
-
-  // Make sure the working variables are within the new limits.
-  if (prevControllerOutput_ > 1) {
-    prevControllerOutput_ = 1;
-  } else if (prevControllerOutput_ < 0) {
-    prevControllerOutput_ = 0;
-  }
-
-  outMin_  = outMin;
-  outMax_  = outMax;
-  outSpan_ = outMax - outMin;
-}
-
-void PID::setTunings(float Kc, float tauI, float tauD) {
-  // Verify that the tunings make sense.
-  if (Kc == 0.0 || tauI < 0.0 || tauD < 0.0) {
-    return;
-  }
-
-  // Store raw values to hand back to user on request.
-  pParam_ = Kc;
-  iParam_ = tauI;
-  dParam_ = tauD;
-
-  float tempTauR;
-
-  if (tauI == 0.0) {
-    tempTauR = 0.0;
-  } else {
-    tempTauR = (1.0 / tauI) * tSample_.count();
-  }
-
-  // For "bumpless transfer" we need to rescale the accumulated error.
-  if (inAuto) {
-    if (tempTauR == 0.0) {
-      accError_ = 0.0;
-    } else {
-      accError_ *= (Kc_ * tauR_) / (Kc * tempTauR);
-    }
-  }
-
-  Kc_   = Kc;
-  tauR_ = tempTauR;
-  tauD_ = tauD / tSample_.count();
+mbed_error_status_t PID::updateIntegralGain(uint32_t i) {
+  m_IGain.store(i);
+  return MBED_SUCCESS;
 }
 
-void PID::reset(void) {
-  float scaledBias = 0.0;
-
-  if (usingFeedForward) {
-    scaledBias = (bias_ - outMin_) / outSpan_;
-  } else {
-    scaledBias = (realOutput_ - outMin_) / outSpan_;
-  }
-
-  prevControllerOutput_ = scaledBias;
-  prevProcessVariable_  = (processVariable_ - inMin_) / inSpan_;
-
-  // Clear any error in the integral.
-  accError_ = 0;
+mbed_error_status_t PID::updateDerivativeGain(uint32_t d) {
+  m_DGain.store(d);
+  return MBED_SUCCESS;
 }
 
-void PID::setMode(int mode) {
-  // We were in manual, and we just got set to auto.
-  // Reset the controller internals.
-  if (mode != 0 && !inAuto) {
-    reset();
-  }
-
-  inAuto = (mode != 0);
+mbed_error_status_t PID::updateDeadzone(float deadzone) {
+  m_deadzone.store(deadzone);
+  return MBED_SUCCESS;
 }
 
-void PID::setInterval(std::chrono::duration<float> interval) {
-  if (interval > 0s) {
-    // Convert the time-based tunings to reflect this change.
-    tauR_ *= (interval / tSample_);
-    accError_ *= (tSample_ / interval);
-    tauD_ *= (interval / tSample_);
-    tSample_ = interval;
-  }
+uint32_t PID::reportProportionalGain() const {
+  return m_PGain.load();
 }
 
-void PID::setSetPoint(float sp) {
-  setPoint_ = sp;
+uint32_t PID::reportIntegralGain() const {
+  return m_IGain.load();
 }
 
-void PID::setProcessValue(float pv) {
-  processVariable_ = pv;
+uint32_t PID::reportDerivativeGain() const {
+  return m_DGain.load();
 }
 
-void PID::setBias(float bias) {
-  bias_            = bias;
-  usingFeedForward = 1;
+float PID::reportDeadzone() const {
+  return m_deadzone.load();
 }
 
-void PID::setDeadZoneError(float error) {
-  deadZoneError_ = error;
+void PID::reset() {
+  m_IPath.store(0);
+  m_pastError.store(0);
+  m_pastPV.store(0);
+  m_timerMutex.lock();
```

Comment:
typically its preferable to use RAII:
https://en.cppreference.com/w/cpp/thread/scoped_lock